docs/source/miscellaneous/doc/adr/0014-cassandra-targeted-sweep.md

```diff
@@ -0,0 +1,283 @@
+
+- If the commit timestamp is greater than the sweep timestamp, targeted sweep must wait until the sweep timestamp increases
+enough so that the entry can be processed. This is generally not an issue since it is only likely to happen when targeted
+sweep is processing writes that were written within the last hour.
+
+#### Targeted Sweep Implementation
+
+Targeted Sweep has a number of background threads (as controlled the install config) continuously cycling through the shards for
+each strategy. To find the next shard to sweep, the thread requests a TimeLock lock for the shard and strategy. If successful, it
+starts an iteration of targeted sweep; otherwise, it requests a lock for the next shard, finally giving up and pausing if it cycles
+unsuccessfully through all the shards. This mechanism ensures synchronization across multiple nodes of the service. Assuming the
+thread successfully acquired a lock, we do the following
+  1. Calculate the *sweep timestamp* for the sweep strategy used and read the *last swept timestamp* for the shard and strategy
+    from *sweepProgressPerShard*.
+  2. Get a batch of `WriteInfo`s from the sweep queue: read from the sweep queue as described above, where *minTs* is the *last
+     swept timestamp* and *maxTs* is the *sweep timestamp*.
+        - If we do not find any candidates, skip to step 6.
+  3. For each of the start timestamps in the batch, we check if the transaction was committed; if not we must abort it (this is the
+  same behaviour as legacy sweep). If a transaction was committed, but **after** the *sweep timestamp*, we must not progress
+  targeted sweep past its start timestamp, so we remove it and all the writes with greater start timestamps from the batch.
+  4. Delete all writes that are referenced to from aborted transactions.
+  5. Partition the remaining `WriteInfo`s by `Cell` and `TableReference`, and then take the greatest start timestamp from each
+    partition. We only need one ranged tombstone per partition here, as all the other writes to that cell and table in this batch
+    are going to have a lower start timestamp and are therefore going to be deleted as well.
+       - If the strategy is conservative: write a garbage collection sentinel, then put a ranged tombstone deleting all versions of that
+         cell between 0 and write timestamp - 1, not deleting the sentinel or the write.
+       - If the strategy is thorough
+            - If the write was a tombstone, then put a ranged tombstone deleting all versions of that cell between -1 and write timestamp,
+              deleting both a potentially existing sentinel and the write.
+            - Otherwise, put a ranged tombstone deleting all versions of that cell between -1 and write timestamp - 1, deleting a
+              potentially existing sentinel, but not the write.
+  6. If the new sweep progress described in greater detail below guarantees that the minimum start timestamp swept in the future
```

Comment:
(described in greater detail below)

Reply:
Is step 6 async or synchronous with the other steps?

Reply:
No -- it should not be done before we do the actual deletes, and should not be done after persisting progress as we could end up with zombies if we die after progressing.

Reply:
I guess as long as we're efficiently deleting the entire rows, the overhead of doing it synchronously shouldn't matter.